core/src/main/java/com/google/errorprone/bugpatterns/EmptyCatch.java

```diff
@@ -0,0 +1,372 @@
+import static com.sun.source.tree.Tree.Kind.CATCH;
+import static com.sun.source.tree.Tree.Kind.BLOCK;
+import static com.sun.source.tree.Tree.Kind.IF;
+import static com.sun.source.tree.Tree.Kind.FOR_LOOP;
+import static com.sun.source.tree.Tree.Kind.DO_WHILE_LOOP;
+import static com.sun.source.tree.Tree.Kind.WHILE_LOOP;
+import static com.sun.source.tree.Tree.Kind.ENHANCED_FOR_LOOP;
+import static com.sun.source.tree.Tree.Kind.SWITCH;
+import static com.sun.source.tree.Tree.Kind.TRY;
+import static com.sun.source.tree.Tree.Kind.VARIABLE;
+
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.CatchTree;
+import com.sun.source.tree.StatementTree;
+import com.sun.source.tree.TryTree;
+import com.sun.source.tree.BlockTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.ExpressionStatementTree;
+
+import com.sun.source.util.TreePath;
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+
+import java.util.List;
+
+/**
+ * A bug checker for empty exception handling block:
+ *
+ * <pre>
+ * {@code
+ * try {
+ *   // do something
+ * } catch (XX e) {
+ *   // empty, only comment, or only log printing statements
+ * }
+ * }
+ * </pre>
+ * We also implement a variety of false positive suppression heuristics.
+ * <p>
+ * For more detail, refer to the paper:
+ * "Simple Testing Can Prevent Most Critical Failures: 
+ *  An Analysis of Production Failures in Distributed Data-intensive Systems"
+ *  Yuan et al. Proceedings of the 11th USENIX Symposium on Operating Systems Design 
+ *  and Implementation (OSDI), 2014
+ *
+ * @author yuan@eecg.utoronto.ca (Ding Yuan)
+ */
+@BugPattern(name = "EmptyCatch",
+    summary = "Exception handling block is empty",
+    explanation = "It is rarely correct to do nothing in response to a caught exception. In production systems, "
+            + "Murphy's law often applies: Anything that can go wrong, will go wrong. "
+            + "Therefore it is particularly important to properly handle the exceptions "
+            + "regardless how rarely they are likely to occur.\n\n"
+            + "Read Effective Java, Item 65, \"Don't ignore exceptions\"; "
+            + "[Google Java Style guide section 6.2] "
+            + "(https://google.github.io/styleguide/javaguide.html#s6.2-caught-exceptions); and "
+            + "\"[Simple Testing Can Prevent Most Critical Failures] "
+            + "(http://www.eecg.toronto.edu/~yuan/papers/failure_analysis_osdi14.pdf)\" "
+            + "for more detailed discussions on the harm of swallowing the exceptions. ",
+    category = JDK, maturity = EXPERIMENTAL, severity = WARNING)
+public class EmptyCatch extends BugChecker implements TryTreeMatcher {
+  /* If one of the following exceptions is ignored, we suppress the warning. */
+  ImmutableSet<String> harmlessExceptions = ImmutableSet.of("java.io.FileNotFoundException", 
+    "java.lang.ReflectiveOperationException", "java.lang.ClassNotFoundException",
```

Comment:
This is a great point. Now thinking about it they should not be necessary. We initially came up with the heuristic b/c of the code snippets like the following (both examples are from HBase):
```java
       try {
         return fs.getFileStatus(path);
       } catch (FileNotFoundException e) {
         // Try another file location
       }
```
Or 
```java
htd = null;
try {
   htd = get(FSUtils.getTableName(d));
 } catch (FileNotFoundException fnfe) {
   // inability of retrieving one HTD shouldn't stop getting the remaining
   LOG.warn("Trouble retrieving htd", fnfe);
 }
if (htd == null) continue;
```
Later we came up with the general idea to suppress the false positives if the exception is handled via divergence in control- or data-flow outside of the catch block (the handledByControlflow and handledByDataflow methods). Now the above cases will suppressed by control-flow and data-flow checks respectively. I checked our evaluation results on the 10 systems and indeed those false positives caused by these exceptions can all be covered by handledByControlflow and handledByDataflow. Therefore these harmlessExceptions should no longer be useful and they probably will cause false negatives.